src/cache_cf.cc

```diff
@@ -311,139 +312,219 @@ skip_ws(const char* s)
+/// octal base per Parser::Tokenizer::int64())
+static int64_t
+EvalNumber(const SBuf &raw)
+{
+    auto numberParser = Parser::Tokenizer(raw);
+    int64_t result = 0;
+    if (!numberParser.int64(result, 0, true))
+        throw TextException(ToSBuf("malformed integer near ", raw), Here());
+    RejectTrailingGarbage("integer", raw, numberParser);
+    return result;
 }
 
+/// IsIfStatementOpening() helper that interprets input prefix as a preprocessor condition
 static bool
-StrToInt(const char* str, long& number)
+EvalBoolExpr(Parser::Tokenizer &tk)
 {
-    assert(str != nullptr);
+    const auto operand = ExtractOperand("preprocessor condition", tk);
+
+    static const auto keywordTrue = SBuf("true");
+    if (operand == keywordTrue)
+        return true;
 
-    char* end;
-    number = strtol(str, &end, 0);
+    static const auto keywordFalse = SBuf("false");
+    if (operand == keywordFalse)
+        return false;
+
+    const auto lhs = operand;
+
+    const auto op = ExtractOperator("equality sign in an equality condition", tk);
+    static const auto keywordEqual = SBuf("=");
+    if (op != keywordEqual)
+        throw TextException(ToSBuf("expected equality sign (=) but got ", op), Here());
+
+    const auto rhs = ExtractOperand("right-hand operand of an equality condition", tk);
+    return EvalNumber(lhs) == EvalNumber(rhs);
+}
+
+/// interprets input as the first line of a preprocessor `if` statement
+/// \returns std::nullopt if input does not look like an `if` statement
+/// \returns `if` condition value if input is an `if` statement
+static std::optional<bool>
+IsIfStatementOpening(Parser::Tokenizer tk)
+{
+    // grammar: space* "if" space condition space* END
+    (void)SkipOptionalSpace(tk);
+    const auto keywordIf = SBuf("if");
+    if (tk.skip(keywordIf) && SkipOptionalSpace(tk)) {
+        const auto condition = tk.remaining();
+        const auto result = EvalBoolExpr(tk);
+        (void)SkipOptionalSpace(tk);
+        RejectTrailingGarbage("preprocessor condition", condition, tk);
+        return result;
+    }
 
-    return (end != str) && (*end == '\0'); // returns true if string contains nothing except number
+    // e.g., "iffy_error_responses on"
+    return std::nullopt;
 }
 
+/// interprets input as an `else` or `endif` line of a preprocessor `if` statement
+/// \returns false if input does not look like an `else` or `endif` line
 static bool
-EvalBoolExpr(const char* expr)
+IsIfStatementLine(const SBuf &keyword, Parser::Tokenizer tk)
 {
-    assert(expr != nullptr);
-    if (strcmp(expr, "true") == 0) {
-        return true;
-    } else if (strcmp(expr, "false") == 0) {
-        return false;
-    } else if (const char* equation = strchr(expr, '=')) {
-        const char* rvalue = skip_ws(equation + 1);
-        char* lvalue = (char*)xmalloc(equation - expr + 1);
-        xstrncpy(lvalue, expr, equation - expr + 1);
-        trim_trailing_ws(lvalue);
+    // grammar: space* keyword space* END
```

Comment:
Hmm. right. The wrapper function naming and code structure makes it seem like any skipping done causes a rejection.

The problem this is the opposite. A line **without* spaces after the found token is **not** rejected as garbage. For example;
 * the old parser (as already mentioned) accepted "else if" and treated it as an else-statement with a nested if-statement
 * this logic will accept "elseif" and do the same.
.. both of which are unintentional and broken/undefined behaviour.

Reply:
> A line *_without_ spaces after the found token is **not** rejected as garbage.

That statement is misleading in this context because such a line is **not** accepted either: A line with garbage after letters matching a given keyword K results in IsIfStatementLine(K) returning false. That is correct behavior for that function and the grammar it defines/supports -- preprocessor keyword K was not found. After IsIfStatementLine(K) returns false, other parsing code (possibly including other IsIfStatementLine(X) calls) deals with that line, rejecting it as needed.

> For example; ... this logic will accept "elseif" and [treat it as an else-statement with a nested if-statement]

That example is incorrect or misleading. PR code does not accept "elseif". It correctly rejects "elseif", first as not a preprocessor line and then as an unknown squid.conf directive:

```
2024/11/05 10:51:40| ERROR: unrecognized directive near 'elseif'
2024/11/05 10:51:41| FATAL: configuration failure: Found 1 unrecognized directive(s)
```

If we add a directive named like `elsewhere_services ...` or `iffy_clients_enabled ...`, PR preprocessor code will not need to be modified for Squid to correctly recognize that new directive despite its name starting with the same letters as the preprocessor keyword "else" or "if". Needless to say, the chances of us adding those exact directives are pretty much zero, but I am using them to illustrate that PR code structure is correct and can support future changes such as addition of new directives (or new preprocessor keywords that are more likely to match some directive name prefixes!).